lib/spack/spack/package.py

```diff
@@ -2195,6 +2193,58 @@ def possible_dependencies(*pkg_or_spec, **kwargs):
     return visited
 
 
+#: Registers which are the detectable packages, by repo and package name
+detectable_packages = collections.defaultdict(list)
+
+
+def detectable(decorated_cls):
+    """Mark a package class as detectable and add a default implementation
+    for detection logic.
+
+    The decorator registers the class as detectable and attaches to it a
+    default implementation of ``determine_spec_details``.
+
+    Args:
+        decorated_cls: class to be decorated
+
+    Returns:
+        Decorated class
+    """
+    @classmethod
+    def determine_spec_details(cls, prefix, exes_in_prefix):
+        exes_by_version = collections.defaultdict(list)
+        # The default filter function is the identity function for the
+        # list of executables
+        filter_fn = getattr(cls, 'filter_detected_exes', lambda x, exes: exes)
+        exes_in_prefix = filter_fn(prefix, exes_in_prefix)
+        for exe in exes_in_prefix:
+            version_str = cls.determine_version(exe)
+            if version_str:
+                exes_by_version[version_str].append(exe)
+
+        specs = []
+        for version_str, exes in exes_by_version.items():
+            specs.extend(cls.determine_variants(exes, version_str))
+
+        return sorted(specs)
+
+    @classmethod
+    def determine_variants(cls, exes, version_str):
```

Comment:
Needs a docstring. It's not clear to me what this function does. If it's only something you override to find variants, why does it need access to the version_str? I think `determine_spec_details` should be the function that creates a `Spec` class, and things like `determine_version` and `determine_variants` should be called if they are defined and overridden as necessary.

Reply:
> If it's only something you override to find variants, why does it need access to the version_str?

Because the default workflow in `determine_spec_details` is currently to first group by version and then determine one or more set of variants for each version group.

Reply:
Regarding this comment in the PR summary above:

> However, I don't like that determine_version returns a version while determine_variants returns a package name, version, and variants (full spec). I think determine_variants should only return a string of variants to add to the spec in whatever function calls it.

I completely get your point and I'll try to simplify the interface. The rationale for returning multiple specs here is that once you determined a version, you can inspect all the executables marked as version X and figure out:
1. How many specs they represent (with their additional details)
2. Which extra attributes they need

Note that for a given `prefix` and `version` there might be more than one spec - think for instance if on the system I had installed cross-compilers along with "normal" compilers and they are at the same version.

Reply:
I understand that the same prefix may have multiple versions, or multiple specs at the same version, but each exe should only have a single spec, correct?

Reply:
> I understand that the same prefix may have multiple versions, or multiple specs at the same version, but each exe should only have a single spec, correct?

Sure, but you might need the other exes in the prefix to determine it correctly. Think `cmake` and `ccmake`: `cmake` must be there, and if you have `ccmake` it means `cmake+ncurses`. Compilers are another example for the `languages=...` part of the spec.